File: 102. Binary Tree Level Order Traversal/binary_tree_level_order_traversal.py
# Recursive
# T.C.->O(n)
# S.C.->O(n)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def traverse(self, node, level, ans):
        if not node:
            return
        if len(ans) == level:
            ans.append([])  # Start a new level
        ans[level].append(node.val)
        self.traverse(node.left, level + 1, ans)
        self.traverse(node.right, level + 1, ans)
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        self.traverse(root, 0, ans)
        return ans


# Iterative
# T.C.->O(n)
# S.C.->O(n)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        ans = []
        if root is None:
            return ans
        q = deque()
        q.append(root)
        while q:
            size = len(q)
            level = []
            for _ in range(size):
                node = q.popleft()
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
                level.append(node.val)
            ans.append(level)

        return ans

File: 102. Binary Tree Level Order Traversal/binary_tree_level_order_traversal.cpp
// Recursive
// T.C.->O(n)
// S.C.->O(n)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traverse(TreeNode* node, int level, vector<vector<int>>& result) {
        if (!node) return;
        
        // If we're entering a new level, create a new vector
        if (result.size() == level)
            result.push_back(vector<int>());
        
        result[level].push_back(node->val);

        traverse(node->left, level + 1, result);
        traverse(node->right, level + 1, result);
    }

    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        traverse(root, 0, result);
        return result;
    }
};

// Iterative
// T.C.->O(n)
// S.C.->O(n)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(root== NULL)
            return ans;
        queue<TreeNode*>q;
        q.push(root);
        while(!q.empty())
        {
            int size=q.size();
            vector<int> level;
            for(int i=0;i<size;i++)
            {
                TreeNode* node=q.front();
                q.pop();
                if(node->left!=NULL)
                    q.push(node->left);
                if(node->right!=NULL)
                    q.push(node->right);
                level.push_back(node->val);
            }
            ans.push_back(level);
        }
        return ans;
    }
};
